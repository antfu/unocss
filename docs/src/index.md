---
layout: home
title: "UnoCSS: The instant on-demand Atomic CSS engine"

hero:
  name: "UnoCSS"
  text: "The instant on-demand Atomic CSS engine"
  actions:
    - theme: brand
      text: Get started
      link: /presets
    - theme: alt
      text: Play
      link: /playground

features:
  - title: Fully customizable
    details: No core utilities, all functionalities are provided via presets.
  - title: Instant
    details: No parsing, no AST, no scanning. Itâ€™s 5x faster than Windi CSS or Tailwind JIT.
  - title: Lightweight
    details: "Zero deps and browser friendly: ~6kb min+brotli."
---

<Home />