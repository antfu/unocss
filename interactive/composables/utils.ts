import { uniq } from '@unocss/core'
import type { ResultItem, RuleItem } from '~/types'
import { mdnIndex } from '~/data/mdn-index'

export function extractColors(css: string) {
  return Array.from(css.matchAll(/\brgba?\((.+?)\)/g))
    .map((i) => {
      const [r, g, b, a] = i[1].split(',').map(i => parseInt(i.trim()))
      if (Number.isNaN(r))
        return ''
      if (!Number.isNaN(a))
        return `rgba(${r}, ${g}, ${b}, ${a})`
      return `rgb(${r}, ${g}, ${b})`
    })
    .filter(Boolean)
}

export function getFeatureUsage(css: string) {
  const props = uniq([...css.matchAll(/^\s+(\w[\w-]+)\:/mg)].map(i => i[1]))
  const functions = uniq([...css.matchAll(/\b(\w+)\(/mg)].map(i => `${i[1]}()`))
  return [...props, ...functions]
    .filter(i => mdnIndex.find(s => s.title === i))
}

export function getDocs(item: RuleItem) {
  return item.features?.map(i => mdnIndex.find(s => s.title === i)!) || []
}

export function getUtilsOfFeature(name: string) {
  return [...(featuresMap.get(name) || [])]
}

export function findAlias(item: RuleItem) {
  return [...matchedMap.values()].filter(i => i.body === item.body && i.class !== item.class)
}

export function goRandom() {
  const keys = Array.from(matchedMap.keys())
  const index = Math.round(keys.length * Math.random())
  useRouter().push({ query: { s: keys[index] } })
}

export function sampleArray<T>(arr: T[], count: number) {
  return Array.from({ length: count }, _ => arr[Math.round(Math.random() * (arr.length - 1))])
}

export function getItemId(item: ResultItem) {
  if (item.type !== 'rule')
    return `${item.type}:${item.title}`
  return item.class
}
