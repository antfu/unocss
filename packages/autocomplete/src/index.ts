export * from './create'
export * from './types'
export * from './parse'
export * from './utils'
