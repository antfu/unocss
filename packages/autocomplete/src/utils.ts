export function searchUsageBoundary(line: string, index: number) {
  let start = index
  let end = index

  const regex = /[^\s>"'`;]/
  while (start && regex.test(line.charAt(start - 1))) --start
  while (end < line.length && regex.test(line.charAt(end))) ++end

  return {
    content: line.slice(start, end),
    start,
    end,
  }
}
