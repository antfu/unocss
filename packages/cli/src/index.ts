import { existsSync, promises as fs } from 'fs'
import { basename, dirname, relative, resolve } from 'pathe'
import fg from 'fast-glob'
import consola from 'consola'
import { cyan, dim, green } from 'colorette'
import { debounce } from 'perfect-debounce'
import { createGenerator, toArray } from '@unocss/core'
import { loadConfig } from '@unocss/config'
import { version } from '../package.json'
import { PrettyError, handleError } from './errors'
import { defaultConfig } from './config'
import type { CliOptions, ResolvedCliOptions } from './types'

const name = 'unocss'

export async function resolveOptions(options: CliOptions) {
  if (!options.patterns?.length) {
    throw new PrettyError(
      `No glob patterns, try ${cyan(`${name} <path/to/**/*>`)}`,
    )
  }

  return options as ResolvedCliOptions
}

export async function build(_options: CliOptions) {
  const fileCache = new Map<string, string>()

  const cwd = _options.cwd || process.cwd()
  const options = await resolveOptions(_options)
  const { config, sources: configSources } = await loadConfig(cwd, options.config)

  const uno = createGenerator(
    config,
    defaultConfig,
  )

  const files = await fg(options.patterns, { cwd, absolute: true })
  await Promise.all(
    files.map(async (file) => {
      fileCache.set(file, await fs.readFile(file, 'utf8'))
    }),
  )

  consola.log(green(`${name} v${version}`))
  consola.start(`UnoCSS ${options.watch ? 'in watch mode...' : 'for production...'}`)

  const debouncedBuild = debounce(
    async () => {
      generate(options).catch(handleError)
    },
    100,
  )

  const startWatcher = async () => {
    if (!options.watch)
      return

    const { watch } = await import('chokidar')
    const { patterns } = options
    const ignored = ['**/{.git,node_modules}/**']

    const watcher = watch(patterns, {
      ignoreInitial: true,
      ignorePermissionErrors: true,
      ignored,
      cwd,
    })

    if (configSources.length)
      watcher.add(configSources)

    watcher.on('all', async (type, file) => {
      const absolutePath = resolve(cwd, file)

      if (configSources.includes(absolutePath)) {
        uno.setConfig((await loadConfig()).config)
        consola.info(`${cyan(basename(file))} changed, setting new config`)
      }
      else {
        consola.log(`${green(type)} ${dim(file)}`)

        if (type.startsWith('unlink'))
          fileCache.delete(absolutePath)
        else
          fileCache.set(absolutePath, await fs.readFile(absolutePath, 'utf8'))
      }

      debouncedBuild()
    })

    consola.info(
      `Watching for changes in ${
        toArray(patterns)
          .map(i => cyan(i))
          .join(', ')}`,
    )
  }

  await generate(options)

  startWatcher().catch(handleError)

  async function generate(options: ResolvedCliOptions) {
    const outFile = resolve(options.cwd || process.cwd(), options.outFile ?? 'uno.css')
    const { css, matched } = await uno.generate(
      [...fileCache].join('\n'),
      {
        preflights: options.preflights,
        minify: options.minify,
      },
    )

    const dir = dirname(outFile)
    if (!existsSync(dir))
      await fs.mkdir(dir, { recursive: true })
    await fs.writeFile(outFile, css, 'utf-8')

    if (!options.watch) {
      consola.success(
      `${[...matched].length} utilities generated to ${cyan(
        relative(process.cwd(), outFile),
      )}\n`,
      )
    }
  }
}
