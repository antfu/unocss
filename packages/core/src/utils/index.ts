export * from './escape'
export * from './object'
export * from './basic'
export * from './helpers'
export * from './map'
export * from './set'
export * from './layer'
export * from './variantGroup'
export * from './warn'
export * from './handlers'
