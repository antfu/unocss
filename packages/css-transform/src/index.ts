import { notNull } from '@unocss/core'
import type { RuleMeta, StringifiedUtil, UnoGenerator } from '@unocss/core'
import type { CssNode, ListItem, Selector, SelectorList, StyleSheet } from 'css-tree'
import { List, clone, generate, parse, walk } from 'css-tree'

type Writeable<T> = { -readonly [P in keyof T]: T[P] }

export async function transformCSS(css: string, uno: UnoGenerator, filename?: string) {
  if (!css.includes('@apply'))
    return css

  const ast = parse(css, {
    parseAtrulePrelude: false,
    positions: true,
    filename,
  })

  if (ast.type !== 'StyleSheet')
    return css

  const stack: Promise<void>[] = []

  const processNode = async(node: CssNode, item: ListItem<CssNode>, list: List<CssNode>) => {
    if (node.type !== 'Rule')
      return

    await Promise.all(
      node.block.children.map(async(childNode, childItem) => {
        if (!(childNode.type === 'Atrule' && childNode.name === 'apply' && childNode.prelude))
          return

        const raw = '-'
        const context = uno.makeContext(raw, [raw, raw, []])

        let classNames: string[] = []

        if (childNode.prelude.type === 'AtrulePrelude') {
          classNames = childNode.prelude.children
            .map(node => node.type === 'Identifier' ? node.name : null)
            .filter(notNull).toArray()
        }

        if (childNode.prelude.type === 'Raw')
          classNames = childNode.prelude.value.split(/\s+/g)

        // expand shortcuts
        const expanded = classNames.map((i) => {
          return uno.expandShortcut(i, context) || ([[i], undefined] as [string[], RuleMeta | undefined])
        }).filter(notNull)

        if (!expanded.length)
          return

        const parentSelector = generate(node.prelude)

        const utils = (
          await Promise.all(
            expanded.map(async i => await uno.stringifyShortcuts([raw, raw, []], context, i[0], i[1])),
          ))
          .filter(notNull).flat()
          .sort((a, b) => a[0] - b[0])
          .reduce((acc, item) => {
            const target = acc.find(i => i[1] === item[1])
            if (target)
              target[2] += item[2]
            else
              acc.push(item as Writeable<StringifiedUtil>)
            return acc
          }, [] as Writeable<StringifiedUtil>[])

        for (const i of utils) {
          if (i[3]) {
            const newNodeCss = `${i[3]}{${parentSelector}{${i[2]}}}`
            const insertNodeAst = parse(newNodeCss) as StyleSheet

            list.insertList(insertNodeAst.children, item)
          }
          else if (i[1] && i[1] !== '.\\-') {
            const pseudoClassSelectors = (
              parse(i[1], {
                context: 'selector',
              }) as Selector)
              .children
              .filter(i => i.type === 'PseudoClassSelector')

            const parentSelectorAst = clone(node.prelude) as SelectorList

            parentSelectorAst.children.forEach((i) => {
              if (i.type === 'Selector')
                i.children.appendList(pseudoClassSelectors)
            })

            const newNodeCss = `${generate(parentSelectorAst)}{${i[2]}}`
            const insertNodeAst = parse(newNodeCss) as StyleSheet

            list.insertList(insertNodeAst.children, item)
          }
          else {
            const rules = new List<string>()
              .fromArray(i[2]
                .replace(/;$/, '')
                .split(';'),
              ).map(i => parse(i, {
                context: 'declaration',
              }))

            node.block.children.insertList(rules, childItem)
          }
        }
        node.block.children.remove(childItem)
      }).toArray(),
    )
  }

  walk(ast, (...args) => stack.push(processNode(...args)))

  await Promise.all(stack)

  // @todo merge rules with same media query

  return generate(ast)
}
