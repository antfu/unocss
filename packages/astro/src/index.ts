import { join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import type { AstroIntegration } from 'astro'
import type { VitePluginConfig } from '@unocss/vite'
import VitePlugin from '@unocss/vite'
import type { UserConfigDefaults } from '@unocss/core'
import type { Plugin, ResolvedConfig } from 'vite'
import { RESOLVED_ID_RE } from '../../shared-integration/src/layers'

const UNO_INJECT_ID = 'uno-astro'
const astroCSSKeyRE = /(\?|\&)lang\.css/

interface AstroVitePluginOptions {
  injects: string[]
  injectReset: boolean
}

function AstroVitePlugin(options: AstroVitePluginOptions): Plugin {
  const { injects, injectReset } = options
  const resetInjectPath = injectReset ? injects[0] : undefined
  let command: ResolvedConfig['command']
  let root: string
  let resetCSSInjected = false
  const resolveCSSQueue = new Set<() => void>()

  return {
    name: 'unocss:astro',
    enforce: 'pre',
    configResolved(config) {
      root = config.root
      command = config.command
    },
    async resolveId(id, importer) {
      if (RESOLVED_ID_RE.test(id)) {
        // https://github.com/withastro/astro/blob/087270c61fd5c91ddd37db5c8fd93a8a0ef41f94/packages/astro/src/core/util.ts#L91-L93
        // Align data-astro-dev-id with data-vite-dev-id to fix https://github.com/unocss/unocss/issues/2513
        return this.resolve(join(root, id), importer, { skipSelf: true })
      }

      if (id === UNO_INJECT_ID) {
        if (injectReset) {
          /**
           * When running here, means that this is a new file.
           * We need to make sure that the reset css for each file
           * needs to be loaded first.
           */
          resetCSSInjected = false
        }
        return id
      }

      if (
        injectReset && command === 'serve'
        // css need to be injected after reset style
        && astroCSSKeyRE.test(id) && !resetCSSInjected
      )
        return new Promise(resolve => resolveCSSQueue.add(() => resolve()))

      if (importer?.endsWith(UNO_INJECT_ID) && command === 'serve') {
        const resolved = await this.resolve(id, importer, { skipSelf: true })
        if (resolved) {
          const fsPath = resolved.id

          if (injectReset) {
            if (resetInjectPath!.includes(id)) {
              // Make sure the reset style is injected first
              setTimeout(() => {
                resolveCSSQueue.forEach((res) => {
                  res()
                  resolveCSSQueue.delete(res)
                })
              })
              resetCSSInjected = true
            }
            // css need to be injected after reset style
            else if (id.includes('.css') && !resetCSSInjected) {
              return new Promise((resolve) => {
                resolveCSSQueue.add(() => {
                  resolve(fsPath)
                })
              })
            }
          }

          return fsPath
        }
      }
    },
    load(id) {
      if (id.endsWith(UNO_INJECT_ID))
        return injects.join('\n')
    },
  }
}

export interface AstroIntegrationConfig<Theme extends object = object> extends VitePluginConfig<Theme> {
  /**
   * Include reset styles
   * When passing `true`, `@unocss/reset/tailwind.css` will be used
   * @default false
   */
  injectReset?: string | boolean

  /**
   * Inject UnoCSS entry import for every astro page
   * @default true
   */
  injectEntry?: boolean | string

  /**
   * Inject extra imports for every astro page
   * @default []
   */
  injectExtra?: string[]
}

export default function UnoCSSAstroIntegration<Theme extends object>(
  options: AstroIntegrationConfig<Theme> = {},
  defaults?: UserConfigDefaults,
): AstroIntegration {
  const {
    injectEntry = true,
    injectReset = false,
    injectExtra = [],
  } = options

  return {
    name: 'unocss',
    hooks: {
      'astro:config:setup': async ({ config, updateConfig, injectScript }) => {
        // Adding components to UnoCSS's extra content
        const source = resolve(fileURLToPath(config.srcDir), 'components/**/*').replace(/\\/g, '/')
        options.content ||= {}
        options.content.filesystem ||= []
        options.content.filesystem.push(source)

        const injects: string[] = []
        if (injectReset) {
          const resetPath = typeof injectReset === 'string'
            ? injectReset
            : '@unocss/reset/tailwind.css'
          injects.push(`import ${JSON.stringify(resetPath)}`)
        }
        if (injectEntry) {
          injects.push(typeof injectEntry === 'string'
            ? injectEntry
            : 'import "uno.css"')
        }
        if (injectExtra.length > 0)
          injects.push(...injectExtra)

        updateConfig({
          vite: {
            plugins: [AstroVitePlugin({
              injectReset: !!injectReset,
              injects,
            }), ...VitePlugin(options, defaults)],
          },
        })

        if (injects?.length)
          injectScript('page-ssr', `import ${JSON.stringify(UNO_INJECT_ID)}`)
      },
    },
  }
}
