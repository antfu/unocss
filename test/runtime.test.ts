import { createGenerator } from '@unocss/core'
import presetUno from '@unocss/preset-uno'
import presetTagify from '@unocss/preset-tagify'
import { describe, expect, test } from 'vitest'
import { autoPrefixer } from '../packages/runtime/src/utils'

function mockElementWithStyle() {
  const store: any = {}
  return {
    style: {
      WebkitFilter: '',
      WebkitPerspective: '',
      WebkitVars: '',
      msHyphens: '',
      MsNonHyphens: '',
      setProperty(key: string, val: string) {
        store[key] = val
      },
      getPropertyValue(key: string) {
        return store[key]
      },
    },
  }
}

const targets = [
  'filter', // whitelisted in function
  'perspect-1px', // Webkit prefix
  'hyphens-auto', // ms prefix
  '[--vars:value]', // css variable
  '[-vars:value]',
  '[vars:value]',
  '[filter:none]',
  '[perspective:0]',
  '[hyphens:none]',
  '[non-hyphens:none]',
].join(' ')

describe('runtime auto prefixer', () => {
  test('without autoprfixer', async () => {
    const uno = createGenerator({
      presets: [
        presetUno(),
      ],
    })

    const { css } = await uno.generate(targets)
    expect(css).toMatchSnapshot()
  })

  test('using autoprfixer', async () => {
    const uno = createGenerator({
      presets: [
        presetUno(),
      ],
      postprocess: [
        autoPrefixer(mockElementWithStyle().style as any),
      ],
    })

    const { css } = await uno.generate(targets)
    expect(css).toMatchSnapshot()
  })

  test('runtime tagify', async () => {
    const uno = createGenerator({
      presets: [
        presetUno(),
        presetTagify(),
      ],
    })

    const { css } = await uno.generate(`
      <flex>
        <text-red> red text </text-red>
        <text-green5:10 />
        <m-1 class="p2"> margin </m-1>
        <btn> shortcut </btn>
        <hover:color-red> variant </hover:color-red>
        <scale-2> zoom </scale-2>
        <shadow-xl> modal </shadow-xl>
      </flex>
    `)
    expect(css).toMatchSnapshot()
  })
})
