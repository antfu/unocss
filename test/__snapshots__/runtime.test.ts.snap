// Vitest Snapshot v1

exports[`runtime auto prefixer > runtime tagify 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x:var(--un-empty,/*!*/ /*!*/);--un-pan-y:var(--un-empty,/*!*/ /*!*/);--un-pinch-zoom:var(--un-empty,/*!*/ /*!*/);--un-scroll-snap-strictness:proximity;--un-ordinal:var(--un-empty,/*!*/ /*!*/);--un-slashed-zero:var(--un-empty,/*!*/ /*!*/);--un-numeric-figure:var(--un-empty,/*!*/ /*!*/);--un-numeric-spacing:var(--un-empty,/*!*/ /*!*/);--un-numeric-fraction:var(--un-empty,/*!*/ /*!*/);--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 #0000;--un-ring-shadow:0 0 #0000;--un-shadow-inset:var(--un-empty,/*!*/ /*!*/);--un-shadow:0 0 #0000;--un-ring-inset:var(--un-empty,/*!*/ /*!*/);--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur:var(--un-empty,/*!*/ /*!*/);--un-brightness:var(--un-empty,/*!*/ /*!*/);--un-contrast:var(--un-empty,/*!*/ /*!*/);--un-drop-shadow:var(--un-empty,/*!*/ /*!*/);--un-grayscale:var(--un-empty,/*!*/ /*!*/);--un-hue-rotate:var(--un-empty,/*!*/ /*!*/);--un-invert:var(--un-empty,/*!*/ /*!*/);--un-saturate:var(--un-empty,/*!*/ /*!*/);--un-sepia:var(--un-empty,/*!*/ /*!*/);--un-backdrop-blur:var(--un-empty,/*!*/ /*!*/);--un-backdrop-brightness:var(--un-empty,/*!*/ /*!*/);--un-backdrop-contrast:var(--un-empty,/*!*/ /*!*/);--un-backdrop-grayscale:var(--un-empty,/*!*/ /*!*/);--un-backdrop-hue-rotate:var(--un-empty,/*!*/ /*!*/);--un-backdrop-invert:var(--un-empty,/*!*/ /*!*/);--un-backdrop-opacity:var(--un-empty,/*!*/ /*!*/);--un-backdrop-saturate:var(--un-empty,/*!*/ /*!*/);--un-backdrop-sepia:var(--un-empty,/*!*/ /*!*/);}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x:var(--un-empty,/*!*/ /*!*/);--un-pan-y:var(--un-empty,/*!*/ /*!*/);--un-pinch-zoom:var(--un-empty,/*!*/ /*!*/);--un-scroll-snap-strictness:proximity;--un-ordinal:var(--un-empty,/*!*/ /*!*/);--un-slashed-zero:var(--un-empty,/*!*/ /*!*/);--un-numeric-figure:var(--un-empty,/*!*/ /*!*/);--un-numeric-spacing:var(--un-empty,/*!*/ /*!*/);--un-numeric-fraction:var(--un-empty,/*!*/ /*!*/);--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 #0000;--un-ring-shadow:0 0 #0000;--un-shadow-inset:var(--un-empty,/*!*/ /*!*/);--un-shadow:0 0 #0000;--un-ring-inset:var(--un-empty,/*!*/ /*!*/);--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur:var(--un-empty,/*!*/ /*!*/);--un-brightness:var(--un-empty,/*!*/ /*!*/);--un-contrast:var(--un-empty,/*!*/ /*!*/);--un-drop-shadow:var(--un-empty,/*!*/ /*!*/);--un-grayscale:var(--un-empty,/*!*/ /*!*/);--un-hue-rotate:var(--un-empty,/*!*/ /*!*/);--un-invert:var(--un-empty,/*!*/ /*!*/);--un-saturate:var(--un-empty,/*!*/ /*!*/);--un-sepia:var(--un-empty,/*!*/ /*!*/);--un-backdrop-blur:var(--un-empty,/*!*/ /*!*/);--un-backdrop-brightness:var(--un-empty,/*!*/ /*!*/);--un-backdrop-contrast:var(--un-empty,/*!*/ /*!*/);--un-backdrop-grayscale:var(--un-empty,/*!*/ /*!*/);--un-backdrop-hue-rotate:var(--un-empty,/*!*/ /*!*/);--un-backdrop-invert:var(--un-empty,/*!*/ /*!*/);--un-backdrop-opacity:var(--un-empty,/*!*/ /*!*/);--un-backdrop-saturate:var(--un-empty,/*!*/ /*!*/);--un-backdrop-sepia:var(--un-empty,/*!*/ /*!*/);}
/* layer: default */
m-1{margin:0.25rem;}
flex{display:flex;}
scale-2{--un-scale-x:0.02;--un-scale-y:0.02;transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
.p2{padding:0.5rem;}
hover\\\\:color-red:hover,
text-red{--un-text-opacity:1;color:rgba(248,113,113,var(--un-text-opacity));}
text-green5\\\\:10{color:rgba(34,197,94,0.1);}
shadow-xl{--un-shadow:var(--un-shadow-inset) 0 20px 25px -5px var(--un-shadow-color, rgba(0,0,0,0.1)),var(--un-shadow-inset) 0 8px 10px -6px var(--un-shadow-color, rgba(0,0,0,0.1));box-shadow:var(--un-ring-offset-shadow, 0 0 #0000), var(--un-ring-shadow, 0 0 #0000), var(--un-shadow);}"
`;

exports[`runtime auto prefixer > using autoprefixer 1`] = `
"/* layer: default */
.\\\\[--vars\\\\:value\\\\]{--vars:value;}
.\\\\[-vars\\\\:value\\\\],
.\\\\[vars\\\\:value\\\\]{-webkit-vars:value;}
.\\\\[filter\\\\:none\\\\]{-webkit-filter:none;}
.\\\\[hyphens\\\\:none\\\\]{-ms-hyphens:none;}
.\\\\[non-hyphens\\\\:none\\\\]{non-hyphens:none;}
.\\\\[perspective\\\\:0\\\\]{-webkit-perspective:0;}
.perspect-1px{-webkit-perspective:1px;}
.hyphens-auto{-webkit-hyphens:auto;-ms-hyphens:auto;}
.filter{-webkit-filter:var(--un-blur) var(--un-brightness) var(--un-contrast) var(--un-drop-shadow) var(--un-grayscale) var(--un-hue-rotate) var(--un-invert) var(--un-saturate) var(--un-sepia);}"
`;

exports[`runtime auto prefixer > without autoprefixer 1`] = `
"/* layer: default */
.\\\\[--vars\\\\:value\\\\]{--vars:value;}
.\\\\[-vars\\\\:value\\\\]{-vars:value;}
.\\\\[filter\\\\:none\\\\]{filter:none;}
.\\\\[hyphens\\\\:none\\\\]{hyphens:none;}
.\\\\[non-hyphens\\\\:none\\\\]{non-hyphens:none;}
.\\\\[perspective\\\\:0\\\\]{perspective:0;}
.\\\\[vars\\\\:value\\\\]{vars:value;}
.perspect-1px{-webkit-perspective:1px;perspective:1px;}
.hyphens-auto{-webkit-hyphens:auto;-ms-hyphens:auto;hyphens:auto;}
.filter{filter:var(--un-blur) var(--un-brightness) var(--un-contrast) var(--un-drop-shadow) var(--un-grayscale) var(--un-hue-rotate) var(--un-invert) var(--un-saturate) var(--un-sepia);}"
`;
